apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx

```diff
@@ -0,0 +1,405 @@
+  const [questionContent, setQuestionContent] = useState<VNode | null>(null);
+
+  // Used if the question receives a number or string response
+  const [userInputResponse, setUserInputResponse] = useState('');
+
+  // Used if the question has a checkbox or multiple choice response
+  const [selectedOptions, setSelectedOptions] = useState<Set<string>>(() => new Set<string>());
+
+  const [placeholder, setPlaceholder] = useState('');
+  const [answerLabel, setAnswerLabel] = useState<string | null>(null);
+
+  const [grade, setGrade] = useState<number | null>(null);
+
+  const handleSelectOption = (option: string) => {
+    if (prompt.answerType === 'radio') {
+      // The user can only select one option.
+      setSelectedOptions(new Set([option]));
+      return;
+    }
+    setSelectedOptions((prevSelectedOptions) => {
+      const newSelectedOptions = new Set(prevSelectedOptions);
+      if (newSelectedOptions.has(option)) {
+        newSelectedOptions.delete(option);
+      } else {
+        newSelectedOptions.add(option);
+      }
+      return newSelectedOptions;
+    });
+  };
+
+  const handleClearSelectedOptions = () => {
+    setSelectedOptions(new Set<string>());
+  };
+
+  const handleGenerateNewVariant = async () => {
+    // Clear the previous question content
+    setQuestionContent(null);
+    // Ensure the component is rendered before typesetting MathJax
+    // TODO: This results in flickering. Find a better solution.
+    await new Promise((resolve) => requestAnimationFrame(resolve));
+
+    // Clear the grade shown to the user
+    setGrade(null);
+
+    // Clear the user input response
+    setUserInputResponse('');
+
+    // Clear the user's selected options
+    handleClearSelectedOptions();
+
+    // Generate a new question variant
+    const questionVariant = generateSampleQuestionVariant(prompt.id);
+    setVariant(questionVariant);
+
+    // Split the question into different parts: regular text, MathJax, and bold text.
+    // MathJax is delimited by $$...$$ or $...$
+    // Bold text is delimited by **...**.
+    const questionParts = questionVariant.question
+      .split(/(\$\$[\s\S]+?\$\$|\$[\s\S]+?\$|\*\*[\s\S]+?\*\*)/g)
+      .filter(Boolean)
+      .map((part, index) => ({
+        index,
+        content: part,
+      }));
+
+    setQuestionContent(
+      <span>
+        {questionParts.map((part) => {
+          // Bold text
+          if (part.content.startsWith('**') && part.content.endsWith('**')) {
+            return (
+              <strong key={`question-content-${part.index}`}>{part.content.slice(2, -2)}</strong>
+            );
+          }
+
+          // MathJax
+          if (
+            (part.content.startsWith('$$') && part.content.endsWith('$$')) ||
+            (part.content.startsWith('$') && part.content.endsWith('$'))
+          ) {
+            return <span key={`question-content-${part.index}`}>{part.content}</span>;
```

Comment:
This is the other piece I commented on above. Ensuring the key changes when the MathJax does ensures we get a brand-new DOM node on which we can perform type-setting.

```suggestion
            return <span key={`${part.index}-${part.content}`}>{part.content}</span>;
```

Reply:
I made this change in conjunction with rendering the question content `span` tags without a state variable:

https://github.com/PrairieLearn/PrairieLearn/blob/a62658cd3aa4443d80fba66dcefe92e716b219fe/apps/prairielearn/src/ee/pages/instructorAiGenerateDrafts/SampleQuestionDemo.tsx#L221-L242